Завести на бэкенде сущность Issue (Feature/21)

artsaban opened the pull request:
Добавил класс Issue и IssueController для обработки запросов

Завёл 2 сущности Issue для тестов

Commit:
Add Issue and IssueController

Card #21

Commit:
Changes in IssueController

Card #21

Commit:
GET request support

Card #21

Commit:
Change type of status field

Card #21

Commit:
Status enum moved into Issue class

Card #21

artsaban commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Как разработчику бэкенда, мне необходима сущность Issue, чтобы реализовать для неё endpoints](https://trello.com/c/ET4Fw1tf/21-issue-endpoints)

Commit:
Some small changes in .class-files

Card #21

Commit:
hashCounter uses in constructor now

Card #21

Commit:
Change first arguments in HashMap' .put method

Card #21

kevromster commented:
Общие замечания по commit messages, на будущее:
- Напоминаю статью: http://chris.beams.io/posts/git-commit/
- "Сhanges in IssueController" и т.п. - плохо, та как непонятно, что за изменения и зачем. Пишем, что действительно сделано в этом коммите. Если надо, добавляем подробный body.
- Ссылку на номер задачи пишем в формате Card: #X (во всех коммитах пропущено двоеточие)

kevromster commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/Issue.java:
```diff
@@ -0,0 +1,55 @@
+    private String description;
+    private Status status;
+
+    public Issue() {
```
Этот конструктор зачем? У нас нет бизнес-понятия "пустой ишуй"

artsaban replied:
Если использовать @RequestBody то он вроде как нужен был. Убрать?

kevromster commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,43 @@
+@RequestMapping(value = {"/issue", "/issue/{hash}"})
+public class IssueController {
+
+    private static Map<Long, Issue> issueMap = new HashMap<Long, Issue>();
```
Напиши здесь комментарий в духе // TODO: temporary issues storage for testing purposes. Remove when implement access to real issues through CLI core.

artsaban replied:
Done

kevromster commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,43 @@
+    private static long hashCounter;
+
+    static {
+
```
не очень понимаю, зачем здесь пустая строка

artsaban replied:
Done

kevromster commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,43 @@
+public class IssueController {
+
+    private static Map<Long, Issue> issueMap = new HashMap<Long, Issue>();
+    private static long hashCounter;
```
эта переменная, кажется, не нужна

artsaban replied:
Я ее в конструкторе использую. Хотя можно убрать и просто явно подавать числа типа лонг.

kevromster commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,43 @@
+
+    @GetMapping
+    public ResponseEntity<Issue> get(@PathVariable("hash") long hash) {
+        return ResponseEntity.status(HttpStatus.OK).body(issueMap.get(hash));
```
Если ишуя нет, всё равно вернёшь OK - непорядок

artsaban replied:
Done

Commit:
Remove empty constructor

Card: #21

Commit:
Remove unused variable hashCount, fixed GET request status

If we try to put element form HashMap and element with this
key doesn't exist returns 404 Not Found status

Card: #21

artsaban commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Как разраб фронтенда, хочу GET-эндпойнт для отдельной таски](https://trello.com/c/JYrKqV8W/6-get)

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/Issue.java:
```diff
@@ -0,0 +1,52 @@
+        this.status = status;
+    }
+
+    enum Status {
```
Этому `enum` лучше бы быть `public`, а то получается, что из других пакетов можно звать getStatus(), но с этим совершенно ничего нельзя поделать.

artsaban replied:
Done

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<Issue> get(@PathVariable("hash") long hash) {
```
А почему бы нам не вернуть просто `Issue` здесь? Имхо `ResponseEntity` не нужно, ведь мы возвращаем `Issue` чисто по смыслу.
Алсо метод лучше назвать `getIssue`, мне кажется. Это будет более говорящее имя.

artsaban replied:
А если возвращать Issue, то статус при этом можно возвращать?

kevromster replied:
а на отсутствующий ишуй просто null вернуть?

artsaban replied:
Done

mgtriffid replied:
@whyme7 , да, статус будет 200 OK если ишью вернётся. 
@kevromster , а вот `null` возвращать не рекомендую. Можно завести `IssueNotFoundException` рядышком, бросить его, а в контроллере объявить метод, аннотированный 

```
@ExceptionHandler(IssueNotFoundException.class)
@ResponseStatus(HttpStatus.NOT_FOUND)
```

, и тогда если твой `getIssue` бросает `IssueNotFoundException`, то клиент получит 404 Not Found.
Описано тут: https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc#errors-and-rest

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Map;
+
+@RestController
+@RequestMapping(value = {"/issue", "/issue/{hash}"})
```
Вот какой-то [гайд по именованию REST путей](http://www.restapitutorial.com/lessons/restfulresourcenaming.html), который советует назвать наш путь `/issues`, а не `/issue`.

artsaban replied:
Done

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Map;
+
+@RestController
+@RequestMapping(value = {"/issue", "/issue/{hash}"})
```
А зачем два пути? Мне кажется, лучше будет перенести вторую часть на метод `getIssue`, мол, там написать `@GetMapping("/{hash}")`. Потому что вдруг мы в этом же контроллере захотим создать `getAllIssues`, тогда его мы оставим с просто аннотацией `@GetMapping` и будет органично смотреться.

artsaban replied:
Done

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<Issue> get(@PathVariable("hash") long hash) {
```
Коммент не на сейчас, а на будущее. Кажется, круто будет делать get не по хэшу, а по taskId. В трелло мы этой радости лишены, а во вменяемых трекерах обычно ссылка просто содержит номер задачи и это кайфово и понятно, и в закладки легко добавить.
Впрочем, так как трекер распределённый, возможно, мы этой радости в принципе лишены. Короче этот коммент скорее мысли вслух, не надо ничего предпринимать сейчас :)

artsaban replied:
Принято :)

Commit:
enum Status is bublic now

Card: #21

Commit:
Change mapping paths, get() renamed, getIssue() returns Issue now

Second path in @RequestMapping removed, get() renamed in getIssue().
getIssue() have own mapping /{hash}.

Card: #21

Commit:
Delete unsed imports

Card: #21

kevromster commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,43 @@
+    @GetMapping(value = "/{hash}")
+    public Issue getIssue(@PathVariable("hash") long hash) {
+        if (!issueMap.containsKey(hash)) {
+            return null;
```
Нужно сделать как @mgtriffid описал тут https://github.com/edu-xored/tracker-web/pull/4#discussion_r84621630

artsaban replied:
Done.

Commit:
getIssue() throws custom IssueNotFoundException if issueMap not include necessary key

Card: #21

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,49 @@
+
+    @ResponseStatus(value = HttpStatus.NOT_FOUND, reason = "Issue not found")
+    public class IssueNotFoundException extends RuntimeException {
+        public IssueNotFoundException() {
```
По-моему, дефолтный конструктор не нужен, он создаётся по умолчанию. А в остальном порядок.
Или у Spring такое требование, чтобы конструктор объявляли, и без него что-то идёт не так?

artsaban replied:
Да, можно и без него. Вроде работает :)
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/com/edu/xored/tracker/IssueController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @ResponseStatus(value = HttpStatus.NOT_FOUND, reason = "Issue not found")
+    public class IssueNotFoundException extends RuntimeException {
```
Ой, а зачем исключение публичное? Этого требует Spring? Если сделать `private class`, будет работать? Если будет - давайте так и сделаем, не нужно давать кому попало бросать наше исключение из нашего уютненького контроллера.

artsaban replied:
Работает. Сделал IssueNotFoundException приватным.
Done.

Commit:
Delete constructor whithout params if IssueNotFoundException class, also IssueNotFoundException private now

Card: #21

mgtriffid reviewed (approved):
LGTM
